SimPEG/electromagnetics/frequency_domain/simulation.py

```diff
@@ -240,6 +240,86 @@ def Jtvec(self, m, v, f=None):
 
         return mkvc(Jtv)
 
+    def getJ(self, m, f=None):
+        """
+        Method to form full J given a model m
+
+        :param numpy.ndarray m: inversion model (nP,)
+        :param SimPEG.electromagnetics.frequency_domain.fields.FieldsFDEM u: fields object
+        :rtype: numpy.ndarray
+        :return: J (ndata, nP)
+        """
+        
+        if getattr(self, "_Jmatrix", None) is None:
+            if f is None:
+                f = self.fields(m)
+
+            Ainv = self.Ainv
+            m_size = self.model.size
+
+            Jmatrix = np.zeros((self.survey.nD, m_size))
+
+            block_count = 0
+            for A_i, freq in zip(Ainv, self.survey.frequencies):
+
+                for src in self.survey.get_sources_by_frequency(freq):
+                    
+                    df_duT, df_dmT = [], []
+                    u_src = f[src, self._solutionType]
+
+                    for rx in src.receiver_list:
+                        
+                        v = np.ones(rx.nD, dtype=float)
+
+                        df_duT, df_dmT = rx.evalDeriv(
+                            src, self.mesh, f, v=v, adjoint=True
+                        )
+
+                        df_duT = np.vstack(df_duT)
+                        ATinvdf_duT = (A_i * df_duT)
+                        dA_dmT = self.getADeriv(freq, u_src, ATinvdf_duT, adjoint=True)
+                        dRHS_dmT = self.getRHSDeriv(freq, src, ATinvdf_duT, adjoint=True)
+                        du_dmT = -dA_dmT
+                        
+                        if not isinstance(dRHS_dmT, Zero):
+                            du_dmT += dRHS_dmT
+                        if not isinstance(df_dmT[0], Zero):
+                            du_dmT += np.hstack(df_dmT)
+
+                        block = np.array(du_dmT, dtype=complex).real.T
+                        Jmatrix[block_count:(block_count + rx.nD)] = block
+                        block_count += rx.nD
```

Comment:
I don't think you can use the "block count" way of doing this to reliably index the matrix for the frequency domain problem.

The `get_sources_by_frequency`  function does not necessarily return the sources in data order, it returns them grouped by common frequency. Normally this isn't an issue for magnetotellurics, however, it's certainly possible to have multiple sources with the same frequency in airborne surveys, and we cannot expect a fixed ordering of them. I think we need to make use of a `data` object from the survey here to get it into the correct rows of `J`.

`data` objects have an `index_dictionary` property that you can use for this. The general way to use it would be:

```
data = Data(self.survey)

for freq in self.survey.frequencies:
    ...
    for src in self.survey.get_sources_by_frequency(freq):
        ...
        for rx in src.receiver_list:
            ...
            row_inds = data.index_dictionary[src][rx]
            J[row_inds] = block
```

Reply:
Thanks @jcapriot , good catch, I'll make use of the `data` object for this.